### PR TITLE
Update to swift-certificates 0.4.0

### PR DESCRIPTION
**Pull request by yim-lee**
- Adjust code
- Re-enable `SigningTests.testCMSCheckCertificateRevocationStatus`
- Implement `SignatureProvider.extractSigningEntity`
- Rework disabling of certificate expiry check
- Update `OCSPPolicy` usage

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -724,7 +724,7 @@ if ProcessInfo.processInfo.environment["SWIFTCI_USE_LOCAL_DEPS"] == nil {
         .package(url: "https://github.com/apple/swift-crypto.git", .upToNextMinor(from: "2.4.0")),
         .package(url: "https://github.com/apple/swift-system.git", .upToNextMinor(from: "1.1.1")),
         .package(url: "https://github.com/apple/swift-collections.git", .upToNextMinor(from: "1.0.1")),
-        .package(url: "https://github.com/apple/swift-certificates.git", .upToNextMinor(from: "0.1.0")),
+        .package(url: "https://github.com/apple/swift-certificates.git", branch: "main"),
```
Comment (yim-lee) [inline Package.swift:727]: Replace with next `swift-certificates` version

**Comment by yim-lee** [comment]
https://github.com/apple/swift/pull/64662

@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

File: Sources/PackageSigning/SignatureProvider.swift [hunk 2]
```diff
@@ -310,6 +308,66 @@ struct CMSSignatureProvider: SignatureProviderProtocol {
             throw SigningError.unableToValidateSignature("\(error)")
         }
     }
+
+    func extractSigningEntity(
```
Comment (yim-lee) [inline Sources/PackageSigning/SignatureProvider.swift:312]: @neonichu Implemented `extractSigningEntity` as part of this PR

Comment (neonichu) [reply Sources/PackageSigning/SignatureProvider.swift:312]: nice, thanks!

**Comment by yim-lee** [comment]
https://github.com/apple/swift/pull/64662

@swift-ci please smoke test

**Comment by yim-lee** [comment]
https://github.com/apple/swift/pull/64662

@swift-ci please test Windows platform

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Review by neonichu — APPROVED** [review]

**Comment by yim-lee** [comment]
macOS cmake build failing with:

```
/Users/ec2-user/jenkins/workspace/swift-package-manager-PR-macos-smoke-test/branch-main/swift-certificates/Sources/X509/CertificatePrivateKey.swift:158:32: error: cannot find type 'SecureEnclave' in scope
        case secureEnclaveP256(SecureEnclave.P256.Signing.PrivateKey)
                               ^~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

Trying to fix with https://github.com/apple/swift-certificates/pull/67, which means we will need another `swift-certificates` release.

File: Sources/PackageSigning/VerifierPolicies.swift [hunk 3]
```diff
@@ -16,99 +16,140 @@ import struct Foundation.Date
...
-                reason: "OCSPVerifierPolicy: leaf certificate \(leaf) has invalid expiry, notValidAfter is earlier than notValidBefore"
-            )
-        }
-        if self.validationTime < leaf.notValidBefore {
-            return .failsToMeetPolicy(reason: "OCSPVerifierPolicy: leaf certificate \(leaf) is not yet valid")
-        }
-        if self.validationTime > leaf.notValidAfter {
-            return .failsToMeetPolicy(reason: "OCSPVerifierPolicy: leaf certificate \(leaf) has expired")
-        }
+    init(failureMode: OCSPFailureMode, httpClient: HTTPClient, validationTime: Date) {
```
Comment (MaxDesiatov) [inline Sources/PackageSigning/VerifierPolicies.swift:133]: nit: could this have a doc comment that clarifies what parameters are for?

**Review by MaxDesiatov — APPROVED** [review]

**Comment by yim-lee** [comment]
Update: to fix cmake build for macOS requires new `swift-crypto` version (https://github.com/apple/swift-crypto/pull/177) and `swift-certificates` (https://github.com/apple/swift-certificates/pull/68).

**Comment by yim-lee** [comment]
https://github.com/apple/swift/pull/64662

@swift-ci please smoke test

**Comment by yim-lee** [comment]
https://github.com/apple/swift/pull/64662

@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please test Windows platform

**Comment by yim-lee** [comment]
https://github.com/apple/swift/pull/64662

@swift-ci please test Windows platform

**Comment by compnerd** [comment]
This breaks the Windows build as the new release of swift-certificates is not portable to non-Darwin platforms.

**Pull request merged**